File: runner/detect.py
"""Capability detection probes for remote hosts.

This module detects host capabilities and platform information to enable
capability-gated rule implementations. Probes are fast, read-only shell
commands that determine what features are available on the target host.

Capability Detection Pattern:
    Each probe is a shell command where exit code 0 = capability present.
    Probes are designed to be:
    - Fast (< 2 seconds)
    - Side-effect free (read-only)
    - Silent (stderr suppressed)

Platform Detection:
    Platform info (family + version) is detected from /etc/os-release with
    fallbacks to /etc/redhat-release and /etc/debian_version. RHEL derivatives
    (Rocky, Alma, CentOS, Oracle Linux) are normalized to family "rhel".

Example:
-------
    >>> from runner.ssh import SSHSession
    >>> from runner.detect import detect_capabilities, detect_platform
    >>>
    >>> with SSHSession("192.168.1.100", user="admin", sudo=True) as ssh:
    ...     caps = detect_capabilities(ssh)
    ...     platform = detect_platform(ssh)
    ...     print(f"Platform: {platform.family} {platform.version}")
    ...     print(f"Has sshd_config.d: {caps['sshd_config_d']}")

"""

from __future__ import annotations

from collections import namedtuple
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from runner.ssh import SSHSession


# ── Platform detection ─────────────────────────────────────────────────────

PlatformInfo = namedtuple("PlatformInfo", ["family", "version"])
"""Platform information for rule filtering.

Attributes:
    family (str): Normalized OS family (e.g., "rhel", "ubuntu", "debian").
    version (int): Major version number.
"""

# OS IDs that are RHEL-compatible and normalized to family "rhel"
RHEL_FAMILY = {"rhel", "centos", "rocky", "almalinux", "ol"}


def detect_platform(ssh: SSHSession) -> PlatformInfo | None:
    """Detect the remote host's OS family and version.

    Uses a fallback chain to detect platform information:
    1. /etc/os-release (preferred, covers most modern distros)
    2. /etc/redhat-release (fallback for older RHEL/CentOS)
    3. /etc/debian_version (fallback for Debian-based systems)

    RHEL derivatives (Rocky, Alma, CentOS, Oracle Linux) are normalized
    to family "rhel" so rules with `family: rhel` match all derivatives.

    Args:
    ----
        ssh: Active SSH session to the target host.

    Returns:
    -------
        PlatformInfo(family, version) on success, None if detection fails.
        Caller should warn and skip platform filtering if None is returned.

    Example:
    -------
        >>> platform = detect_platform(ssh)
        >>> if platform:
        ...     print(f"Detected: {platform.family} {platform.version}")
        ... else:
        ...     print("Warning: Could not detect platform")

    """
    # Try /etc/os-release first (covers most modern distros)
    result = ssh.run("cat /etc/os-release 2>/dev/null")
    if result.ok and result.stdout.strip():
        fields: dict[str, str] = {}
        for line in result.stdout.splitlines():
            if "=" in line:
                k, _, v = line.partition("=")
                fields[k.strip()] = v.strip().strip('"')

        os_id = fields.get("ID", "").lower()
        ver_str = fields.get("VERSION_ID", "0")
        try:
            version = int(ver_str.split(".")[0])
        except (ValueError, IndexError):
            version = 0

        family = "rhel" if os_id in RHEL_FAMILY else os_id
        return PlatformInfo(family=family, version=version)

    # Fallback: /etc/redhat-release for older RHEL/CentOS
    result = ssh.run("cat /etc/redhat-release 2>/dev/null")
    if result.ok and result.stdout.strip():
        import re

        content = result.stdout.strip().lower()
        version = 0
        match = re.search(r"(\d+)", content)
        if match:
            version = int(match.group(1))
        return PlatformInfo(family="rhel", version=version)

    # Fallback: /etc/debian_version for Debian-based
    result = ssh.run("cat /etc/debian_version 2>/dev/null")
    if result.ok and result.stdout.strip():
        try:
            version = int(result.stdout.strip().split(".")[0])
        except (ValueError, IndexError):
            version = 0
        return PlatformInfo(family="debian", version=version)

    return None


# ── Capability probes ──────────────────────────────────────────────────────
#
# Each probe maps a capability name to a shell command.
# Exit code 0 = capability present, non-zero = absent.
#
# Probes must be:
#   - Fast (< 2 seconds)
#   - Side-effect free (read-only)
#   - Silent (suppress stderr with 2>/dev/null)
#
# Adding a new probe:
#   1. Add entry to this dict
#   2. Use capability name in rule `when:` gates
#   3. No code changes needed elsewhere

CAPABILITY_PROBES: dict[str, str] = {
    "sshd_config_d": (
        "test -d /etc/ssh/sshd_config.d && { "
        "grep -qi 'Include.*/etc/ssh/sshd_config.d' /etc/ssh/sshd_config 2>/dev/null || "
        "find /etc/ssh/sshd_config.d -name '*.conf' -print -quit 2>/dev/null | grep -q .; "
        "}"
    ),
    "authselect": "command -v authselect >/dev/null 2>&1 && authselect current >/dev/null 2>&1",
    "authselect_sssd": "authselect current 2>/dev/null | grep -q sssd",
    "crypto_policies": "test -f /etc/crypto-policies/config && command -v update-crypto-policies >/dev/null 2>&1",
    "crypto_policy_modules": (
        "update-crypto-policies --show 2>/dev/null | grep -q ':' || "
        "test -d /usr/share/crypto-policies/policies/modules"
    ),
    "fips_mode": "fips-mode-setup --check 2>/dev/null | grep -qi 'is enabled'",
    "firewalld_nftables": (
        "systemctl is-active firewalld >/dev/null 2>&1 && "
        "firewall-cmd --get-default-zone >/dev/null 2>&1 && "
        "grep -q 'FirewallBackend=nftables' /etc/firewalld/firewalld.conf 2>/dev/null"
    ),
    "firewalld_iptables": (
        "systemctl is-active firewalld >/dev/null 2>&1 && "
        "grep -q 'FirewallBackend=iptables' /etc/firewalld/firewalld.conf 2>/dev/null"
    ),
    "systemd_resolved": "systemctl is-active systemd-resolved >/dev/null 2>&1",
    "pam_faillock": "test -f /etc/security/faillock.conf",
    "grub_bls": "test -d /boot/loader/entries && grep -q 'blscfg' /etc/default/grub 2>/dev/null",
    "grub_legacy": "test -f /boot/grub2/grub.cfg && ! test -d /boot/loader/entries",
    "journald_primary": "systemctl is-active systemd-journald >/dev/null 2>&1",
    "rsyslog_active": "systemctl is-active rsyslog >/dev/null 2>&1",
    "fapolicyd": "rpm -q fapolicyd >/dev/null 2>&1",
    "selinux": "command -v getenforce >/dev/null 2>&1 && [ \"$(getenforce 2>/dev/null)\" != 'Disabled' ]",
    "aide": "rpm -q aide >/dev/null 2>&1",
    "tpm2": "test -c /dev/tpmrm0 || test -c /dev/tpm0",
    "usbguard": "rpm -q usbguard >/dev/null 2>&1",
    "dnf_automatic": "rpm -q dnf-automatic >/dev/null 2>&1",
    "gdm": "rpm -q gdm >/dev/null 2>&1",
    "tmux": "command -v tmux >/dev/null 2>&1",
}


def detect_capabilities(ssh: SSHSession, *, verbose: bool = False) -> dict[str, bool]:
    """Run all capability probes and return a dict of results.

    Executes each probe in CAPABILITY_PROBES and records whether it
    succeeded (exit code 0). Results are used by select_implementation()
    to choose capability-gated rule implementations.

    Args:
    ----
        ssh: Active SSH session to the target host.
        verbose: If True, print debug info for failed probes to stderr.

    Returns:
    -------
        Dict mapping capability name to bool (True if present).

    Example:
    -------
        >>> caps = detect_capabilities(ssh, verbose=True)
        >>> if caps["sshd_config_d"]:
        ...     print("Using drop-in config implementation")
        ... else:
        ...     print("Using main config file implementation")

    """
    caps = {}
    for name, cmd in CAPABILITY_PROBES.items():
        result = ssh.run(cmd)
        caps[name] = result.ok
        if verbose and not result.ok:
            import sys

            print(
                f"    [probe] {name}: exit={result.exit_code} stderr={result.stderr!r}",
                file=sys.stderr,
            )
    return caps

File: internal/detect/detect.go
// Package detect runs SSH probes to determine which capabilities a
// remote host has. Capabilities gate which [api.Implementation] is
// selected from a rule's implementations list.
package detect

import (
	"context"
	"sort"

	"github.com/Hanalyx/kensa/api"
)

// probe is a single capability probe: a name and the shell command to
// run on the target host.
type probe struct {
	name string
	cmd  string
}

// probes is the canonical list of capability probes. Each probe runs
// via [api.Transport.Run]; exit code 0 means the capability is present.
var probes = []probe{
	{
		"sshd_config_d",
		`[ -d /etc/ssh/sshd_config.d ] && grep -qiE '^\s*Include\s+/etc/ssh/sshd_config\.d' /etc/ssh/sshd_config`,
	},
	{
		"authselect",
		`command -v authselect >/dev/null 2>&1`,
	},
	{
		"crypto_policies",
		`command -v update-crypto-policies >/dev/null 2>&1`,
	},
	{
		"fips_mode",
		// RHEL: fips-mode-setup --check. Ubuntu Pro: same binary exists.
		// Universal fallback: /proc/sys/crypto/fips_enabled = 1.
		`fips-mode-setup --check 2>/dev/null | grep -q 'FIPS mode is enabled' || [ "$(cat /proc/sys/crypto/fips_enabled 2>/dev/null)" = "1" ]`,
	},
	{
		"firewalld_nftables",
		`systemctl is-active firewalld >/dev/null 2>&1 && command -v nft >/dev/null 2>&1`,
	},
	{
		"pam_faillock",
		`[ -f /etc/security/faillock.conf ]`,
	},
	{
		"pam_pwquality",
		`[ -f /etc/security/pwquality.conf ]`,
	},
	{
		"grub_bls",
		`[ -d /boot/loader/entries ]`,
	},
	{
		"selinux",
		`getenforce 2>/dev/null | grep -q Enforcing`,
	},
	{
		"aide",
		`command -v aide >/dev/null 2>&1`,
	},
	{
		"fapolicyd",
		`systemctl list-unit-files fapolicyd.service 2>/dev/null | grep -q fapolicyd`,
	},
	{
		"usbguard",
		`systemctl list-unit-files usbguard.service 2>/dev/null | grep -q usbguard`,
	},
	{
		"systemd_resolved",
		`systemctl is-active systemd-resolved >/dev/null 2>&1`,
	},
	{
		"nftables",
		`command -v nft >/dev/null 2>&1`,
	},
	{
		"firewalld",
		`systemctl list-unit-files firewalld.service 2>/dev/null | grep -q firewalld`,
	},
	{
		"rsyslog",
		`systemctl list-unit-files rsyslog.service 2>/dev/null | grep -q rsyslog`,
	},
	{
		"journald",
		`systemctl is-active systemd-journald >/dev/null 2>&1`,
	},
	{
		"auditd",
		`systemctl is-active auditd >/dev/null 2>&1`,
	},
	{
		"cron",
		`systemctl list-unit-files crond.service cron.service 2>/dev/null | grep -qE '(crond|cron)\.service'`,
	},
	{
		"at",
		`command -v at >/dev/null 2>&1`,
	},
	{
		"coredump_systemd",
		`systemctl list-unit-files systemd-coredump.socket 2>/dev/null | grep -q coredump`,
	},
	{
		"sssd",
		`systemctl list-unit-files sssd.service 2>/dev/null | grep -q sssd`,
	},
	{
		"chronyd",
		`systemctl list-unit-files chronyd.service 2>/dev/null | grep -q chronyd`,
	},
	{
		"dnf_automatic",
		`systemctl list-unit-files dnf-automatic.timer 2>/dev/null | grep -q dnf-automatic`,
	},
	{
		"subscription_manager",
		`command -v subscription-manager >/dev/null 2>&1`,
	},

	// ── Ubuntu / Debian-specific probes ──────────────────────────────────

	// dpkg is the primary distro discriminator: present on all Debian-
	// family systems, absent on RHEL/EL. Rules gate Ubuntu implementations
	// on requires: [dpkg].
	{
		"dpkg",
		`command -v dpkg >/dev/null 2>&1`,
	},
	// apt gates package_present/package_absent Ubuntu implementations.
	{
		"apt",
		`command -v apt-get >/dev/null 2>&1`,
	},
	// apparmor is the Ubuntu equivalent of SELinux mandatory access control.
	// Detected independently of selinux so rules can require one or the other.
	{
		"apparmor",
		`aa-status 2>/dev/null | grep -q 'apparmor module is loaded'`,
	},
	// ufw is Ubuntu's front-end to nftables/iptables, analogous to firewalld.
	{
		"ufw",
		`systemctl list-unit-files ufw.service 2>/dev/null | grep -q ufw`,
	},
	// apt_unattended_upgrades is the Ubuntu equivalent of dnf_automatic.
	{
		"apt_unattended_upgrades",
		`dpkg -l unattended-upgrades 2>/dev/null | grep -q '^ii'`,
	},
	// ubuntu_advantage detects Ubuntu Pro / Ubuntu Advantage tooling,
	// analogous to subscription_manager on RHEL.
	{
		"ubuntu_advantage",
		`command -v ua >/dev/null 2>&1 || command -v pro >/dev/null 2>&1`,
	},
	// systemd_dbus is true when (a) systemd is the init system,
	// (b) the system D-Bus socket exists, AND (c) the
	// kensa-systemd-helper binary is installed at the FHS path.
	// Phase 4 D-008. The probe runs as the SSH user (kensa-svc);
	// the helper does the actual privileged D-Bus call via sudo.
	// All three conditions must hold for the agent-mode D-Bus
	// path to be exercisable; if any is false, handlers fall
	// back to their direct-SSH shell-out path.
	{
		"systemd_dbus",
		`[ -S /run/dbus/system_bus_socket ] && [ -x /usr/libexec/kensa-systemd-helper ] && systemctl --version >/dev/null 2>&1`,
	},
}

// Detect runs all capability probes against the host reachable via
// transport and returns the resulting [api.CapabilitySet]. Each
// capability is true when its probe exits with code 0. Transport
// errors on individual probes are suppressed — the capability is
// marked false and detection continues with the remaining probes.
func Detect(ctx context.Context, transport api.Transport) (api.CapabilitySet, error) {
	caps := make(api.CapabilitySet, len(probes))
	for _, p := range probes {
		caps[p.name] = runProbe(ctx, transport, p.cmd)
	}
	return caps, nil
}

// runProbe executes one probe command and returns true when it exits
// with code 0. Any transport error is treated as a negative result so
// that a single failing probe cannot abort the overall detection run.
func runProbe(ctx context.Context, transport api.Transport, cmd string) bool {
	res, err := transport.Run(ctx, cmd)
	if err != nil {
		return false
	}
	return res.ExitCode == 0
}

// KnownCapabilities returns the set of capability names that Detect
// can probe for, sorted alphabetically. Used by the --capability /
// -C flag to validate operator-provided KEY=VALUE pairs against
// the canonical vocabulary.
func KnownCapabilities() []string {
	names := make([]string, 0, len(probes))
	for _, p := range probes {
		names = append(names, p.name)
	}
	sort.Strings(names)
	return names
}

// ApplyOverrides returns a copy of detected with every key in
// overrides set to the override value. Keys present in overrides
// but not in detected are still applied — operators may legitimately
// force a capability that wasn't probed (e.g., behind a fake
// transport in tests). Validation that the keys are in
// KnownCapabilities() is the caller's responsibility.
func ApplyOverrides(detected api.CapabilitySet, overrides api.CapabilitySet) api.CapabilitySet {
	if len(overrides) == 0 {
		return detected
	}
	out := make(api.CapabilitySet, len(detected)+len(overrides))
	for k, v := range detected {
		out[k] = v
	}
	for k, v := range overrides {
		out[k] = v
	}
	return out
}
